File: DateTimeAnalyzer_VB/DateTimeAnalyzer_VB/CodeFixProvider.vb
Imports System.Collections.Immutable
Imports Microsoft.CodeAnalysis.Editing
Imports Microsoft.CodeAnalysis.Rename

<ExportCodeFixProvider(LanguageNames.VisualBasic, Name:=NameOf(DateTimeAnalyzer_VBCodeFixProvider)), [Shared]>
Public Class DateTimeAnalyzer_VBCodeFixProvider
    Inherits CodeFixProvider

    Private Const title As String = "Replace DateTime with DateTimeOffset"

    Public NotOverridable Overrides ReadOnly Property FixableDiagnosticIds As ImmutableArray(Of String)
        Get
            Return ImmutableArray.
                   Create(DateTimeAnalyzer_VBAnalyzer.DiagnosticId)
        End Get
    End Property

    Public NotOverridable Overrides Function GetFixAllProvider() As FixAllProvider
        Return WellKnownFixAllProviders.BatchFixer
    End Function

    Public NotOverridable Overrides _
        Async Function RegisterCodeFixesAsync(context As CodeFixContext) As Task

        'Get the root syntax node for the current document
        Dim root = Await context.Document.
            GetSyntaxRootAsync(context.CancellationToken).
            ConfigureAwait(False)

        'Get a reference to the diagnostic to fix
        Dim diagnostic = context.Diagnostics.First()

        'Get the location in the code 
        'editor for the diagnostic
        Dim diagnosticSpan =
            diagnostic.Location.SourceSpan
        'Find the syntax node on the span 
        'where there Is a squiggle
        Dim node = root.FindNode(context.Span)

        'Register a code action that invokes the fix
        'on the current document
        context.RegisterCodeFix(
        CodeAction.
        Create("Replace with DateTimeOffset",
               Function(c) ReplaceDateTimeAsync(context.Document,
                          node, c),
                          equivalenceKey:=title),
                          diagnostic)
    End Function

    Private Async Function ReplaceDateTimeAsync _
            (document As Document,
             node As SyntaxNode,
             cancellationToken As CancellationToken) _
             As Task(Of Document)

        'Get the root syntax node for the current document
        Dim root = Await document.GetSyntaxRootAsync

        Dim newRoot As SyntaxNode = Nothing

        'Convert the syntax node into the specialized kind
        If TypeOf (node) Is IdentifierNameSyntax Then
            Dim convertedNode = DirectCast(node, IdentifierNameSyntax)

            'Create a new syntax node
            Dim newNode = convertedNode.
                          WithIdentifier(SyntaxFactory.
                          ParseToken("DateTimeOffset")).
                          WithTrailingTrivia(node.
                          GetTrailingTrivia)

            'Create a New root syntax node for the current document
            'replacing the syntax node that has diagnostic with
            'a new syntax node
            newRoot = root.ReplaceNode(node, newNode)
        ElseIf TypeOf (node) Is PredefinedTypeSyntax
            Dim convertedNode = DirectCast(node, PredefinedTypeSyntax)

            Dim newIdentifierName = SyntaxFactory.
                          IdentifierName(SyntaxFactory.
                          ParseToken("DateTimeOffset")).
                          WithTrailingTrivia(node.
                          GetTrailingTrivia)
            newRoot = root.ReplaceNode(node, newIdentifierName)
        End If

        'Generate a new document
        Dim newDocument = document.WithSyntaxRoot(newRoot)
        Return newDocument

    End Function
End Class

File: DateTimeAnalyzer_CS/DateTimeAnalyzer_CS/CodeFixProvider.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Composition;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Rename;
using Microsoft.CodeAnalysis.Text;
using Microsoft.CodeAnalysis.Editing;

namespace DateTimeAnalyzer_CS
{
    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(DateTimeAnalyzer_CSCodeFixProvider)), Shared]
    public class DateTimeAnalyzer_CSCodeFixProvider : CodeFixProvider
    {
        private const string title = "Replace with DateTimeOffset";

        public sealed override ImmutableArray<string> FixableDiagnosticIds
        {
            get { return ImmutableArray.Create(DateTimeAnalyzer_CSAnalyzer.DiagnosticId); }
        }

        public sealed override FixAllProvider GetFixAllProvider()
        {
            return WellKnownFixAllProviders.BatchFixer;
        }

        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
        {
            // Get the root syntax node for the current document
            var root = await context.Document.
                GetSyntaxRootAsync(context.CancellationToken).
                ConfigureAwait(false);

            // Get a reference to the diagnostic to fix
            var diagnostic = context.Diagnostics.First();
            // Get the location in the code editor for the diagnostic
            var diagnosticSpan = diagnostic.Location.SourceSpan;

            // Find the syntax node on the span 
            // where there is a squiggle
            var node = root.FindNode(context.Span);

            // If the syntax node is not an IdentifierName
            // return
            if (node is IdentifierNameSyntax == false)
            {
                return;
            }

            // Register a code action that invokes the fix
            // on the current document
            context.RegisterCodeFix(
            CodeAction.Create(title:title,
                              createChangedDocument: 
                              c=> ReplaceDateTimeAsync(context.Document, 
                              node, c), 
                              equivalenceKey:title), diagnostic);
        }


        private async Task<Document> ReplaceDateTimeAsync(Document document, 
                         SyntaxNode node,
                         CancellationToken cancellationToken)
        {
            // Get the root syntax node for the current document
            var root = await document.GetSyntaxRootAsync();

            // Convert the syntax node into the specialized kind
            var convertedNode = (IdentifierNameSyntax)node;

            // Create a new syntax node
            var newNode = convertedNode?.WithIdentifier(SyntaxFactory.
                          ParseToken("DateTimeOffset")).
                          WithLeadingTrivia(node.GetLeadingTrivia()).
                          WithTrailingTrivia(node.GetTrailingTrivia());

            // Create a new root syntax node for the current document
            // replacing the syntax node that has diagnostic with
            // a new syntax node
            var newRoot = root.ReplaceNode(node, newNode);

            // Generate a new document
            var newDocument = document.WithSyntaxRoot(newRoot);
            return newDocument;
        }
    }
}